Test/git-issues/git-issue-356-errors.dfy

```diff
@@ -0,0 +1,226 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+
+// Check every possible conversion range error among int,char,bv,real,ORDINAL
+
+// int to char
+method Test1(n: int)
+{
+  var ch := n as char;
+}
+
+method Test1b(n: int)
+  requires n < 65536
+{
+  var ch := n as char;
+}
+
+method Test1c(n: int)
+  requires 0 <= n
+{
+  var ch := n as char;
+}
+
+method Test1OK(n: int)
+  requires 0 <= n < 65536
+{
+  var ch := n as char;
+}
+
+
+// int to ORDINAL
+method Test2(n: int) {
+  var o: ORDINAL := n as ORDINAL; // must be non-negative // NO ERROR
+}
+  
+method Test2OK(n: int)
+  requires n >= 0
+{
+  var o: ORDINAL := n as ORDINAL;
+}
+
+// real to ORDINAL  
+method Test3(r: real)
+  requires r.Floor as real == r
+{
+  var o: ORDINAL := r as ORDINAL;
+}
+
+method Test3b(r: real)
+  requires r >= 0.0
+{
+  var o: ORDINAL := r as ORDINAL;
+}
+
+method Test3OK(r: real)
+  requires r.Floor as real == r
+  requires r >= 0.0
+{
+  var o: ORDINAL := r as ORDINAL;
+}
+
+// real to int
+method Test4(r: real)
+{
+  var n: int := r as int;
+}
+
+method Test4OK(r: real)
+  requires r.Floor as real == r
+{
+  var n: int := r as int;
+}
+
+// real to char
```

Comment:
If you'd like, you can write these groups of tests more concisely. For example, here are the Test5... tests (plus one that I added to make sure the `case`s are exhaustive).

```
method Test5(r: real) returns (ch: char) {
  if
  case true =>
    ch := r as real;  // error
  case 0.0 <= r == r.Floor as real =>
    ch := r as real;  // error
  case r < 65536.0 && r == r.Floor as real =>
    ch := r as real;  // error
  case 0.0 <= r < 65536.0 =>
    ch := r as real;  // error
  case 0.0 <= r < 65536.0 && r == r.Floor as real =>
    ch := r as real;
}
```

Reply:
Next time. Also I prefer one error per method, just in case the order is non-deterministic or one hides the other.